docs/src/apalache/fold.md

```diff
@@ -0,0 +1,213 @@
+```
+RECURSIVE Cardinality(_)
+Cardinality(set) == IF set = {}
+                    THEN 0
+                    ELSE LET x == CHOOSE y \in set: TRUE
+                         IN 1 + Cardinality( set \ {x} )
+
+CardinalityFold(set) == LET Count(p,q) == p + 1 \* the value of q, the set element, is irrelevant
+                        IN FoldSet( Count, 0, set )
+```
+
+Notice that, in the case of sets, picking an arbitrary element `x`, to remove from the set at each step, utilizes the `CHOOSE` operator. This is a common trait shared by many operators that implement recursion over sets. 
+Since the introduction of folds, the use of `CHOOSE` in Apalache is heavily discouraged as it is both inefficient, as well as nondeterministic (unlike how `CHOOSE` is defined in TLA+ literature). 
+
+So the third advantage of using folds is the  ability to, almost always, avoid using the `CHOOSE` operator.
+
+The downside of folding, compared to general recursion, is the inability to express non-primitively recursive functions.
+For instance, one cannot define the Ackermann function, as a fold. We find that in most specifications, this is not something the users would want to implement anyway, so in practice, we believe it is almost always better to use fold over recursive functions.
+
+### Examples: The versatility of folds
+
+Here we give some examples of common operators, implemented using folds:
+
+```
+\*  Sum of all values of a set of integers
+Sum(set) == LET Plus(p,q) == p + q IN FoldSet( Plus, 0, set )
+
+\*  Re-implementation of UNION setOfSets
+BigUnion(setOfSets) == LET Union(p,q) == p \union q IN FoldSet( Union, {}, setOfSets )
+
+\*  Re-implementation of SelectSeq
+SelectSeq(seq, Test(_)) == LET CondAppend(s,e) == IF Test(e) THEN Append(s, e) ELSE s
+                           IN FoldSeq( CondAppend, <<>>, seq )
+
+\*  Quantify the elements in S matching the predicate P
+Quantify(S, P(_)) == LET CondCount(p,q) == p + IF P(q) THEN 1 ELSE 0
+                     IN FoldSet( CondCount, 0, S )
+
+\* The set of all values in seq
+Range(seq) == LET AddToSet(S, e) == S \union {e}
+              IN LET Range == FoldSeq( AddToSet, {}, seq )
+
+\* Finds the the value that appears most often in a sequence. Returns elIfEmpty for empty sequences
+Mode(seq, elIfEmpty) == LET ExtRange == Range(seq) \union {elIfEmpty}
+                        IN LET CountElem(countersAndCurrentMode, e) ==
+                             LET counters == countersAndCurrentMode[1]
+                                 currentMode == countersAndCurrentMode[2]
+                             IN LET newCounters == [ counters EXCEPT ![e] == counters[e] + 1 ]
+                                IN IF newCounters[e] > newCounters[currentMode]
+                                   THEN << newCounters, e >>
+                                   ELSE << newCounters, currentMode >>
+                           IN FoldSeq( CountElem, <<[ x \in ExtRange |-> 0 ], elIfEmpty >>, seq )[2]
```

Comment:
I guess, one can write this operator in a much nicer way by using `Range`, `SelectSeq`, and `Len`. Perhaps, this example does not demonstrate the power of `FoldSeq` :-)

Reply:
`ExtRange`, `counters`, `currentMode` are all just convenience variables. This looks, at least to me, like a linear mode-calculation algorithm in a programming language.